In [5]:
from pathlib import Path
import json

import pandas as pd


# Find the main project folder.
# This works whether the notebook starts from the project root
# or from inside the notebooks folder.
current_directory = Path.cwd()

if current_directory.name == "notebooks":
    project_root = current_directory.parent
else:
    project_root = current_directory

raw_data_directory = (
    project_root / "data" / "raw" / "singapore"
)

print("Project root:", project_root)
print("Raw data directory:", raw_data_directory)

Project root: e:\PROJECTS (selfstudy)\expernetic-airbnb-data-engineering
Raw data directory: e:\PROJECTS (selfstudy)\expernetic-airbnb-data-engineering\data\raw\singapore


In [6]:
# Detailed listings data
listings_detailed = pd.read_csv(
    raw_data_directory / "listings.csv.gz",
    low_memory=False
)

# Summary listings data
listings_summary = pd.read_csv(
    raw_data_directory / "listings.csv",
    low_memory=False
)

# Detailed calendar data
calendar = pd.read_csv(
    raw_data_directory / "calendar.csv.gz",
    low_memory=False
)

# Detailed review data, including review text
reviews_detailed = pd.read_csv(
    raw_data_directory / "reviews.csv.gz",
    low_memory=False
)

# Summary review data
reviews_summary = pd.read_csv(
    raw_data_directory / "reviews.csv",
    low_memory=False
)

# Neighbourhood list
neighbourhoods = pd.read_csv(
    raw_data_directory / "neighbourhoods.csv",
    low_memory=False
)

# Neighbourhood GeoJSON boundaries
with open(
    raw_data_directory / "neighbourhoods.geojson",
    mode="r",
    encoding="utf-8"
) as geojson_file:
    neighbourhoods_geojson = json.load(geojson_file)

In [8]:
expected_files = [
    "listings.csv.gz",
    "listings.csv",
    "calendar.csv.gz",
    "reviews.csv.gz",
    "reviews.csv",
    "neighbourhoods.csv",
    "neighbourhoods.geojson"
]

for file_name in expected_files:
    file_path = raw_data_directory / file_name

    if file_path.exists():
        print(f"Found: {file_name}")
    else:
        print(f"Missing: {file_name}")

Found: listings.csv.gz
Found: listings.csv
Found: calendar.csv.gz
Found: reviews.csv.gz
Found: reviews.csv
Found: neighbourhoods.csv
Found: neighbourhoods.geojson


In [9]:
datasets = {
    "Detailed Listings": listings_detailed,
    "Summary Listings": listings_summary,
    "Calendar": calendar,
    "Detailed Reviews": reviews_detailed,
    "Summary Reviews": reviews_summary,
    "Neighbourhoods": neighbourhoods
}

for dataset_name, dataframe in datasets.items():
    print(
        f"{dataset_name}: "
        f"{dataframe.shape[0]:,} rows, "
        f"{dataframe.shape[1]:,} columns"
    )

print(
    "GeoJSON features:",
    len(neighbourhoods_geojson.get("features", []))
)

Detailed Listings: 3,097 rows, 90 columns
Summary Listings: 3,247 rows, 19 columns
Calendar: 1,185,155 rows, 5 columns
Detailed Reviews: 41,265 rows, 6 columns
Summary Reviews: 41,265 rows, 2 columns
Neighbourhoods: 55 rows, 2 columns
GeoJSON features: 55


In [10]:
print("Shape:", listings_detailed.shape)

display(listings_detailed.head())

listings_detailed.info()

Shape: (3097, 90)


,id,listing_url,scrape_id,last_scraped,source,name,description,neighborhood_overview,picture_url,host_id,...,review_scores_communication,review_scores_location,review_scores_value,license,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
0,71609,https://www.airbnb.com/rooms/71609,20260629151954,2026-07-02,previous scrape,Ensuite Room (Room 1 & 2) near EXPO,For 3 rooms.Book room 1&2 and room 4,NaN,https://a0.muscache.com/pictures/24453191/3580...,367042,...,4.78,4.26,4.32,NaN,NaN,5,0,5,0,0.11
1,71896,https://www.airbnb.com/rooms/71896,20260629151954,2026-07-02,previous scrape,B&B Room 1 near Airport & EXPO,NaN,NaN,https://a0.muscache.com/pictures/2440674/ac4f4...,367042,...,4.43,4.17,4.04,NaN,NaN,5,0,5,0,0.13
2,71903,https://www.airbnb.com/rooms/71903,20260629151954,2026-07-02,previous scrape,Room 2-near Airport & EXPO,"Like your own home, 24hrs access.",NaN,https://a0.muscache.com/pictures/568743/7bc623...,367042,...,4.64,4.50,4.36,NaN,NaN,5,0,5,0,0.25
3,294281,https://www.airbnb.com/rooms/294281,20260629151954,2026-07-02,previous scrape,5 mins walk from Newton subway,I have 3 bedrooms in a charming British style...,NaN,https://a0.muscache.com/pictures/4607862/87f42...,1521514,...,4.66,4.52,4.39,NaN,NaN,7,1,6,0,0.75
4,344803,https://www.airbnb.com/rooms/344803,20260629151954,2026-07-02,previous scrape,Budget short stay room near EXPO,Direct bus 24 from airport to this place.Walki...,NaN,https://a0.muscache.com/pictures/4240c88e-4dc2...,367042,...,4.58,4.66,4.44,NaN,NaN,5,0,5,0,0.34


<class 'pandas.DataFrame'>
RangeIndex: 3097 entries, 0 to 3096
Data columns (total 90 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            3097 non-null   int64  
 1   listing_url                                   3097 non-null   str    
 2   scrape_id                                     3097 non-null   int64  
 3   last_scraped                                  3097 non-null   str    
 4   source                                        3097 non-null   str    
 5   name                                          3097 non-null   str    
 6   description                                   2918 non-null   str    
 7   neighborhood_overview                         0 non-null      float64
 8   picture_url                                   3094 non-null   str    
 9   host_id                                       3097 non-null   int64  
 10 

In [11]:
print("Shape:", listings_summary.shape)

display(listings_summary.head())

listings_summary.info()

Shape: (3247, 19)


,id,name,host_id,host_profile_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365,number_of_reviews_ltm,license
0,71609,Ensuite Room (Room 1 & 2) near EXPO,367042.0,1.462517e+18,Belinda,East Region,Tampines,1.34537,103.95887,Private room,NaN,92,19,2020-01-17,0.11,5.0,90,0,NaN
1,71896,B&B Room 1 near Airport & EXPO,367042.0,1.462517e+18,Belinda,East Region,Tampines,1.34754,103.95958,Private room,NaN,92,24,2019-10-13,0.13,5.0,90,0,NaN
2,71903,Room 2-near Airport & EXPO,367042.0,1.462517e+18,Belinda,East Region,Tampines,1.34531,103.96100,Private room,NaN,92,46,2020-01-09,0.25,5.0,90,0,NaN
3,294281,5 mins walk from Newton subway,1521514.0,1.462519e+18,Elizbeth,Central Region,Newton,1.31142,103.83924,Private room,32.0,92,131,2020-01-21,0.75,7.0,364,0,NaN
4,344803,Budget short stay room near EXPO,367042.0,1.462517e+18,Belinda,East Region,Tampines,1.34917,103.96090,Private room,15.0,92,60,2023-08-01,0.34,5.0,364,0,NaN


<class 'pandas.DataFrame'>
RangeIndex: 3247 entries, 0 to 3246
Data columns (total 19 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              3247 non-null   int64  
 1   name                            3247 non-null   str    
 2   host_id                         3097 non-null   float64
 3   host_profile_id                 3028 non-null   float64
 4   host_name                       3038 non-null   str    
 5   neighbourhood_group             3247 non-null   str    
 6   neighbourhood                   3247 non-null   str    
 7   latitude                        3247 non-null   float64
 8   longitude                       3247 non-null   float64
 9   room_type                       3247 non-null   str    
 10  price                           2742 non-null   float64
 11  minimum_nights                  3247 non-null   int64  
 12  number_of_reviews               3247 non-null

In [12]:
print("Shape:", calendar.shape)

display(calendar.head())

calendar.info()

Shape: (1185155, 5)


,listing_id,date,available,minimum_nights,maximum_nights
0,71609,2026-07-02,f,92,1125
1,71609,2026-07-03,t,92,1125
2,71609,2026-07-04,t,92,1125
3,71609,2026-07-05,t,92,1125
4,71609,2026-07-06,t,92,1125


<class 'pandas.DataFrame'>
RangeIndex: 1185155 entries, 0 to 1185154
Data columns (total 5 columns):
 #   Column          Non-Null Count    Dtype
---  ------          --------------    -----
 0   listing_id      1185155 non-null  int64
 1   date            1185155 non-null  str  
 2   available       1185155 non-null  str  
 3   minimum_nights  1185155 non-null  int64
 4   maximum_nights  1185155 non-null  int64
dtypes: int64(3), str(2)
memory usage: 45.2 MB


In [13]:
print("Shape:", reviews_detailed.shape)

display(reviews_detailed.head())

reviews_detailed.info()

Shape: (41265, 6)


,listing_id,id,date,reviewer_id,reviewer_name,comments
0,71609,793880,2011-12-19,1456140,Max,The rooms were clean and tidy. Beds very comfo...
1,71609,1731810,2012-07-17,1804182,Jaiya,Good space and quite an interesting home in a ...
2,71609,2162194,2012-09-01,3113461,Zahra,It was a comfortable place. Belinda was a kind...
3,71609,2190615,2012-09-04,1432123,Helmut,We are four mature age travellers and stayed f...
4,71609,3221837,2013-01-02,2759938,Jack,Belinda is a great host! She helps you wheneve...


<class 'pandas.DataFrame'>
RangeIndex: 41265 entries, 0 to 41264
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   listing_id     41265 non-null  int64
 1   id             41265 non-null  int64
 2   date           41265 non-null  str  
 3   reviewer_id    41265 non-null  int64
 4   reviewer_name  41265 non-null  str  
 5   comments       41248 non-null  str  
dtypes: int64(3), str(3)
memory usage: 1.9 MB


In [14]:
print("Shape:", reviews_summary.shape)

display(reviews_summary.head())

reviews_summary.info()

Shape: (41265, 2)


,listing_id,date
0,71609,2011-12-19
1,71609,2012-07-17
2,71609,2012-09-01
3,71609,2012-09-04
4,71609,2013-01-02


<class 'pandas.DataFrame'>
RangeIndex: 41265 entries, 0 to 41264
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   listing_id  41265 non-null  int64
 1   date        41265 non-null  str  
dtypes: int64(1), str(1)
memory usage: 644.9 KB


In [15]:
print("Shape:", neighbourhoods.shape)

display(neighbourhoods.head())

neighbourhoods.info()

Shape: (55, 2)


,neighbourhood_group,neighbourhood
0,Central Region,Bishan
1,Central Region,Bukit Merah
2,Central Region,Bukit Timah
3,Central Region,Downtown Core
4,Central Region,Geylang


<class 'pandas.DataFrame'>
RangeIndex: 55 entries, 0 to 54
Data columns (total 2 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   neighbourhood_group  55 non-null     str  
 1   neighbourhood        55 non-null     str  
dtypes: str(2)
memory usage: 1012.0 bytes


In [16]:
print(neighbourhoods_geojson.keys())

print(
    "Number of neighbourhood features:",
    len(neighbourhoods_geojson["features"])
)

display(
    neighbourhoods_geojson["features"][0]
)

dict_keys(['type', 'features'])
Number of neighbourhood features: 55


{'type': 'Feature',
 'geometry': {'type': 'MultiPolygon',
  'coordinates': [[[[103.953216, 1.382015],
     [103.953504, 1.381932],
     [103.953553, 1.381891],
     [103.953553, 1.381891],
     [103.953504, 1.381932],
     [103.954216, 1.381783],
     [103.954218, 1.381782],
     [103.954799, 1.381687],
     [103.955003, 1.381667],
     [103.955017, 1.381666],
     [103.955156, 1.381652],
     [103.95523, 1.381644],
     [103.955314, 1.381631],
     [103.955417, 1.381622],
     [103.955492, 1.38162],
     [103.955581, 1.381606],
     [103.955663, 1.381592],
     [103.955753, 1.381581],
     [103.955864, 1.381574],
     [103.955953, 1.381567],
     [103.956033, 1.381547],
     [103.956162, 1.381533],
     [103.956247, 1.381521],
     [103.956338, 1.38151],
     [103.956434, 1.3815],
     [103.956512, 1.381489],
     [103.956585, 1.381489],
     [103.956653, 1.381487],
     [103.956725, 1.381481],
     [103.956735, 1.38148],
     [103.956747, 1.381479],
     [103.956771, 1.381477],
     

In [17]:
def create_basic_profile(
    dataframe: pd.DataFrame
) -> pd.DataFrame:
    profile = pd.DataFrame({
        "column_name": dataframe.columns,
        "data_type": dataframe.dtypes.astype(str).values,
        "row_count": len(dataframe),
        "null_count": dataframe.isna().sum().values,
        "null_percentage": (
            dataframe.isna().mean() * 100
        ).round(2).values,
        "unique_count": dataframe.nunique(
            dropna=False
        ).values
    })

    return profile

In [18]:
listings_profile = create_basic_profile(
    listings_detailed
)

display(listings_profile)

,column_name,data_type,row_count,null_count,null_percentage,unique_count
0,id,int64,3097,0,0.00,3097
1,listing_url,str,3097,0,0.00,3097
2,scrape_id,int64,3097,0,0.00,1
3,last_scraped,str,3097,0,0.00,3
4,source,str,3097,0,0.00,2
...,...,...,...,...,...,...
85,calculated_host_listings_count,int64,3097,0,0.00,34
86,calculated_host_listings_count_entire_homes,int64,3097,0,0.00,24
87,calculated_host_listings_count_private_rooms,int64,3097,0,0.00,27
88,calculated_host_listings_count_shared_rooms,int64,3097,0,0.00,9


In [19]:
calendar_profile = create_basic_profile(calendar)

reviews_profile = create_basic_profile(
    reviews_detailed
)

neighbourhoods_profile = create_basic_profile(
    neighbourhoods
)

In [20]:
dataset_overview = []

for dataset_name, dataframe in datasets.items():
    dataset_overview.append({
        "dataset": dataset_name,
        "rows": dataframe.shape[0],
        "columns": dataframe.shape[1],
        "duplicate_rows": dataframe.duplicated().sum(),
        "total_missing_values": dataframe.isna().sum().sum(),
        "memory_mb": round(
            dataframe.memory_usage(deep=True).sum() / 1024**2,
            2
        )
    })

dataset_overview_df = pd.DataFrame(dataset_overview)

display(dataset_overview_df)

,dataset,rows,columns,duplicate_rows,total_missing_values,memory_mb
0,Detailed Listings,3097,90,0,63211,12.42
1,Summary Listings,3247,19,0,5555,1.61
2,Calendar,1185155,5,0,0,150.32
3,Detailed Reviews,41265,6,0,17,18.46
4,Summary Reviews,41265,2,3165,0,2.64
5,Neighbourhoods,55,2,0,0,0.01


In [21]:
import numpy as np


def profile_dataframe(
    dataframe: pd.DataFrame,
    dataset_name: str
) -> pd.DataFrame:
    profile_rows = []

    for column in dataframe.columns:
        series = dataframe[column]

        non_null_values = series.dropna()

        sample_values = (
            non_null_values
            .astype(str)
            .drop_duplicates()
            .head(3)
            .tolist()
        )

        row = {
            "dataset": dataset_name,
            "column_name": column,
            "data_type": str(series.dtype),
            "row_count": len(dataframe),
            "null_count": int(series.isna().sum()),
            "null_percentage": round(
                series.isna().mean() * 100,
                2
            ),
            "unique_count": int(
                series.nunique(dropna=False)
            ),
            "sample_values": " | ".join(sample_values)
        }

        if pd.api.types.is_numeric_dtype(series):
            row["minimum"] = (
                non_null_values.min()
                if not non_null_values.empty
                else np.nan
            )

            row["maximum"] = (
                non_null_values.max()
                if not non_null_values.empty
                else np.nan
            )

            row["mean"] = (
                round(non_null_values.mean(), 2)
                if not non_null_values.empty
                else np.nan
            )
        else:
            row["minimum"] = None
            row["maximum"] = None
            row["mean"] = None

        profile_rows.append(row)

    return pd.DataFrame(profile_rows)

In [22]:
profiles = {}

for dataset_name, dataframe in datasets.items():
    profile = profile_dataframe(
        dataframe,
        dataset_name
    )

    profiles[dataset_name] = profile

    print(f"\n{dataset_name}")
    display(profile.head(10))


Detailed Listings


,dataset,column_name,data_type,row_count,null_count,null_percentage,unique_count,sample_values,minimum,maximum,mean
0,Detailed Listings,id,int64,3097,0,0.00,3097,71609 | 71896 | 71903,7.160900e+04,1.716878e+18,6.873474e+17
1,Detailed Listings,listing_url,str,3097,0,0.00,3097,https://www.airbnb.com/rooms/71609 | https://w...,NaN,NaN,NaN
2,Detailed Listings,scrape_id,int64,3097,0,0.00,1,20260629151954,2.026063e+13,2.026063e+13,2.026063e+13
3,Detailed Listings,last_scraped,str,3097,0,0.00,3,2026-07-02 | 2026-06-29 | 2026-06-30,NaN,NaN,NaN
4,Detailed Listings,source,str,3097,0,0.00,2,previous scrape | city scrape,NaN,NaN,NaN
5,Detailed Listings,name,str,3097,0,0.00,3002,Ensuite Room (Room 1 & 2) near EXPO | B&B Roo...,NaN,NaN,NaN
6,Detailed Listings,description,str,3097,179,5.78,1946,For 3 rooms.Book room 1&2 and room 4 | Like yo...,NaN,NaN,NaN
7,Detailed Listings,neighborhood_overview,float64,3097,3097,100.00,1,,NaN,NaN,NaN
8,Detailed Listings,picture_url,str,3097,3,0.10,2911,https://a0.muscache.com/pictures/24453191/3580...,NaN,NaN,NaN
9,Detailed Listings,host_id,int64,3097,0,0.00,762,367042 | 1521514 | 1975201,2.182320e+05,1.706906e+18,8.806882e+15



Summary Listings


,dataset,column_name,data_type,row_count,null_count,null_percentage,unique_count,sample_values,minimum,maximum,mean
0,Summary Listings,id,int64,3247,0,0.00,3247,71609 | 71896 | 71903,7.160900e+04,1.716878e+18,7.228868e+17
1,Summary Listings,name,str,3247,0,0.00,3113,Ensuite Room (Room 1 & 2) near EXPO | B&B Roo...,NaN,NaN,NaN
2,Summary Listings,host_id,float64,3247,150,4.62,763,367042.0 | 1521514.0 | 17526618.0,2.182320e+05,1.706906e+18,8.806882e+15
3,Summary Listings,host_profile_id,float64,3247,219,6.74,743,1.462516725725391e+18 | 1.4625186372111946e+18...,1.462513e+18,1.716723e+18,1.470914e+18
4,Summary Listings,host_name,str,3247,209,6.44,627,Belinda | Elizbeth | Royal,NaN,NaN,NaN
5,Summary Listings,neighbourhood_group,str,3247,0,0.00,5,East Region | Central Region | West Region,NaN,NaN,NaN
6,Summary Listings,neighbourhood,str,3247,0,0.00,42,Tampines | Newton | Outram,NaN,NaN,NaN
7,Summary Listings,latitude,float64,3247,0,0.00,2423,1.34537 | 1.34754 | 1.34531,1.222720e+00,1.457802e+00,1.310000e+00
8,Summary Listings,longitude,float64,3247,0,0.00,2475,103.95887 | 103.95958 | 103.961,1.036346e+02,1.039852e+02,1.038500e+02
9,Summary Listings,room_type,str,3247,0,0.00,4,Private room | Hotel room | Entire home/apt,NaN,NaN,NaN



Calendar


,dataset,column_name,data_type,row_count,null_count,null_percentage,unique_count,sample_values,minimum,maximum,mean
0,Calendar,listing_id,int64,1185155,0,0.0,3247,71609 | 71896 | 71903,71609.0,1.716878e+18,7.228868e+17
1,Calendar,date,str,1185155,0,0.0,368,2026-07-02 | 2026-07-03 | 2026-07-04,NaN,NaN,NaN
2,Calendar,available,str,1185155,0,0.0,2,f | t,NaN,NaN,NaN
3,Calendar,minimum_nights,int64,1185155,0,0.0,46,92 | 93 | 730,1.0,7.300000e+02,5.022000e+01
4,Calendar,maximum_nights,int64,1185155,0,0.0,388,1125 | 180 | 730,1.0,9.999000e+03,6.527400e+02



Detailed Reviews


,dataset,column_name,data_type,row_count,null_count,null_percentage,unique_count,sample_values,minimum,maximum,mean
0,Detailed Reviews,listing_id,int64,41265,0,0.00,1839,71609 | 71896 | 71903,71609.0,1.675861e+18,3.324451e+17
1,Detailed Reviews,id,int64,41265,0,0.00,41265,793880 | 1731810 | 2162194,251467.0,1.718692e+18,7.726038e+17
2,Detailed Reviews,date,str,41265,0,0.00,4469,2011-12-19 | 2012-07-17 | 2012-09-01,NaN,NaN,NaN
3,Detailed Reviews,reviewer_id,int64,41265,0,0.00,38902,1456140 | 1804182 | 3113461,14922.0,1.708357e+18,1.383238e+15
4,Detailed Reviews,reviewer_name,str,41265,0,0.00,17990,Max | Jaiya | Zahra,NaN,NaN,NaN
5,Detailed Reviews,comments,str,41265,17,0.04,39345,The rooms were clean and tidy. Beds very comfo...,NaN,NaN,NaN



Summary Reviews


,dataset,column_name,data_type,row_count,null_count,null_percentage,unique_count,sample_values,minimum,maximum,mean
0,Summary Reviews,listing_id,int64,41265,0,0.0,1839,71609 | 71896 | 71903,71609.0,1.675861e+18,3.324451e+17
1,Summary Reviews,date,str,41265,0,0.0,4469,2011-12-19 | 2012-07-17 | 2012-09-01,NaN,NaN,NaN



Neighbourhoods


,dataset,column_name,data_type,row_count,null_count,null_percentage,unique_count,sample_values,minimum,maximum,mean
0,Neighbourhoods,neighbourhood_group,str,55,0,0.0,5,Central Region | East Region | North-East Region,None,None,None
1,Neighbourhoods,neighbourhood,str,55,0,0.0,55,Bishan | Bukit Merah | Bukit Timah,None,None,None


In [23]:
for dataset_name, dataframe in datasets.items():
    missing_summary = (
        dataframe.isna()
        .mean()
        .mul(100)
        .sort_values(ascending=False)
        .reset_index()
    )

    missing_summary.columns = [
        "column_name",
        "null_percentage"
    ]

    missing_summary = missing_summary[
        missing_summary["null_percentage"] > 0
    ]

    print(f"\nTop missing columns: {dataset_name}")

    display(missing_summary.head(15))


Top missing columns: Detailed Listings


,column_name,null_percentage
0,neighborhood_overview,100.000000
1,host_since,100.000000
2,host_response_time,100.000000
3,host_thumbnail_url,100.000000
4,host_acceptance_rate,100.000000
5,host_response_rate,100.000000
6,host_verifications,100.000000
7,neighbourhood,100.000000
8,host_total_listings_count,100.000000
9,host_neighbourhood,100.000000



Top missing columns: Summary Listings


,column_name,null_percentage
0,license,46.381275
1,reviews_per_month,43.363104
2,last_review,43.363104
3,price,15.552818
4,host_profile_id,6.744687
5,host_name,6.436711
6,host_id,4.619649
7,calculated_host_listings_count,4.619649



Top missing columns: Calendar


,column_name,null_percentage



Top missing columns: Detailed Reviews


,column_name,null_percentage
0,comments,0.041197



Top missing columns: Summary Reviews


,column_name,null_percentage



Top missing columns: Neighbourhoods


,column_name,null_percentage


Check duplicate rows and duplicate keys

In [24]:
duplicate_summary = pd.DataFrame({
    "dataset": [
        "Detailed Listings",
        "Summary Listings",
        "Calendar",
        "Detailed Reviews",
        "Summary Reviews",
        "Neighbourhoods"
    ],
    "full_duplicate_rows": [
        listings_detailed.duplicated().sum(),
        listings_summary.duplicated().sum(),
        calendar.duplicated().sum(),
        reviews_detailed.duplicated().sum(),
        reviews_summary.duplicated().sum(),
        neighbourhoods.duplicated().sum()
    ]
})

display(duplicate_summary)

,dataset,full_duplicate_rows
0,Detailed Listings,0
1,Summary Listings,0
2,Calendar,0
3,Detailed Reviews,0
4,Summary Reviews,3165
5,Neighbourhoods,0


In [25]:
key_checks = {
    "Detailed Listings duplicate id": (
        listings_detailed["id"].duplicated().sum()
    ),
    "Summary Listings duplicate id": (
        listings_summary["id"].duplicated().sum()
    ),
    "Calendar duplicate listing_id and date": (
        calendar.duplicated(
            subset=["listing_id", "date"]
        ).sum()
    ),
    "Detailed Reviews duplicate id": (
        reviews_detailed["id"].duplicated().sum()
    )
}

for check_name, result in key_checks.items():
    print(f"{check_name}: {result}")

Detailed Listings duplicate id: 0
Summary Listings duplicate id: 0
Calendar duplicate listing_id and date: 0
Detailed Reviews duplicate id: 0


In [26]:
print(reviews_summary.columns.tolist())

['listing_id', 'date']


In [27]:
reviews_summary.duplicated(
    subset=["listing_id", "date"]
).sum()

np.int64(3165)

Check relationships between datasets

In [28]:
listing_ids = set(
    listings_detailed["id"].dropna()
)

calendar_listing_ids = set(
    calendar["listing_id"].dropna()
)

orphan_calendar_ids = (
    calendar_listing_ids - listing_ids
)

print(
    "Calendar listing IDs not found in listings:",
    len(orphan_calendar_ids)
)

Calendar listing IDs not found in listings: 150


In [29]:
review_listing_ids = set(
    reviews_detailed["listing_id"].dropna()
)

orphan_review_ids = (
    review_listing_ids - listing_ids
)

print(
    "Detailed review listing IDs not found in listings:",
    len(orphan_review_ids)
)

Detailed review listing IDs not found in listings: 103


In [30]:
summary_review_listing_ids = set(
    reviews_summary["listing_id"].dropna()
)

orphan_summary_review_ids = (
    summary_review_listing_ids - listing_ids
)

print(
    "Summary review listing IDs not found in listings:",
    len(orphan_summary_review_ids)
)

Summary review listing IDs not found in listings: 103


In [31]:
detailed_columns = set(
    listings_detailed.columns
)

summary_columns = set(
    listings_summary.columns
)

print(
    "Columns in both files:",
    len(detailed_columns & summary_columns)
)

print(
    "\nOnly in detailed listings:"
)

print(
    sorted(
        detailed_columns - summary_columns
    )
)

print(
    "\nOnly in summary listings:"
)

print(
    sorted(
        summary_columns - detailed_columns
    )
)

Columns in both files: 18

Only in detailed listings:
['accommodates', 'amenities', 'availability_30', 'availability_60', 'availability_90', 'availability_eoy', 'bathrooms', 'bathrooms_text', 'bedrooms', 'beds', 'calculated_host_listings_count_entire_homes', 'calculated_host_listings_count_private_rooms', 'calculated_host_listings_count_shared_rooms', 'calendar_last_scraped', 'calendar_updated', 'description', 'estimated_occupancy_l365d', 'estimated_revenue_l365d', 'first_review', 'has_availability', 'host_about', 'host_acceptance_rate', 'host_has_profile_pic', 'host_identity_verified', 'host_is_superhost', 'host_listings_count', 'host_location', 'host_neighbourhood', 'host_picture_url', 'host_profile_url', 'host_response_rate', 'host_response_time', 'host_since', 'host_thumbnail_url', 'host_total_listings_count', 'host_url', 'host_verifications', 'hosts_time_as_host_months', 'hosts_time_as_host_years', 'hosts_time_as_user_months', 'hosts_time_as_user_years', 'instant_bookable', 'last_

In [32]:
detailed_listing_ids = set(
    listings_detailed["id"].dropna()
)

summary_listing_ids = set(
    listings_summary["id"].dropna()
)

print(
    "IDs only in detailed listings:",
    len(
        detailed_listing_ids -
        summary_listing_ids
    )
)

print(
    "IDs only in summary listings:",
    len(
        summary_listing_ids -
        detailed_listing_ids
    )
)

IDs only in detailed listings: 0
IDs only in summary listings: 150


In [33]:
categorical_columns = [
    "room_type",
    "property_type",
    "neighbourhood_cleansed",
    "host_is_superhost"
]

for column in categorical_columns:
    if column in listings_detailed.columns:
        print(f"\nValue counts for: {column}")

        display(
            listings_detailed[column]
            .value_counts(
                dropna=False
            )
            .head(20)
        )


Value counts for: room_type


room_type
Private room       1720
Entire home/apt    1226
Hotel room           83
Shared room          68
Name: count, dtype: int64


Value counts for: property_type


property_type
Private room in rental unit           633
Entire serviced apartment             625
Room in hotel                         476
Entire rental unit                    396
Private room in home                  154
Private room in condo                 130
Private room in serviced apartment    125
Room in boutique hotel                 93
Entire condo                           84
Room in aparthotel                     53
Entire home                            44
Private room in hostel                 36
Private room in townhouse              33
Private room in bungalow               32
Room in hostel                         28
Shared room in bed and breakfast       24
Shared room in hostel                  18
Private room in bed and breakfast      13
Room in serviced apartment             13
Private room in villa                  12
Name: count, dtype: int64


Value counts for: neighbourhood_cleansed


neighbourhood_cleansed
River Valley       404
Kallang            272
Outram             255
Rochor             246
Novena             188
Bukit Merah        187
Downtown Core      168
Queenstown         159
Geylang            158
Singapore River    145
Bedok              133
Marine Parade       88
Orchard             81
Clementi            71
Newton              54
Hougang             42
Jurong West         40
Tanglin             40
Bukit Timah         39
Jurong East         33
Name: count, dtype: int64


Value counts for: host_is_superhost


host_is_superhost
f      2630
t       408
NaN      59
Name: count, dtype: int64

In [34]:
numerical_columns = [
    "price",
    "accommodates",
    "bedrooms",
    "beds",
    "minimum_nights",
    "maximum_nights",
    "number_of_reviews",
    "review_scores_rating",
    "availability_365",
    "latitude",
    "longitude"
]

available_numerical_columns = [
    column
    for column in numerical_columns
    if column in listings_detailed.columns
]

display(
    listings_detailed[
        available_numerical_columns
    ].describe().T
)

,count,mean,std,min,25%,50%,75%,max
accommodates,3097.0,2.743623,1.806573,1.00000,2.00000,2.000000,4.000000,16.000000
bedrooms,1956.0,1.438139,0.827752,0.00000,1.00000,1.000000,2.000000,7.000000
beds,1290.0,1.695349,1.464149,1.00000,1.00000,1.000000,2.000000,22.000000
minimum_nights,3097.0,52.173071,53.887280,1.00000,1.00000,92.000000,92.000000,730.000000
maximum_nights,3097.0,720.504036,711.812196,1.00000,365.00000,365.000000,1125.000000,9999.000000
number_of_reviews,3097.0,12.481434,52.583910,0.00000,0.00000,1.000000,6.000000,1492.000000
review_scores_rating,1736.0,4.543007,0.620943,1.00000,4.33000,4.750000,5.000000,5.000000
availability_365,3097.0,281.775589,108.977380,0.00000,237.00000,335.000000,363.000000,364.000000
latitude,3097.0,1.308243,0.028059,1.22272,1.29184,1.302630,1.315956,1.457802
longitude,3097.0,103.844772,0.041550,103.63457,103.83634,103.844437,103.860030,103.985176


In [35]:
display(
    listings_detailed["price"]
    .dropna()
    .head(20)
)

3      $31.71
4      $14.66
6      $26.24
8      $17.67
10     $46.23
11     $45.89
17     $18.96
19     $83.37
20     $23.40
21     $31.71
22     $61.89
23     $23.97
24    $176.10
26     $21.75
27     $24.63
29     $40.97
32     $21.68
33     $64.88
34     $47.35
35     $80.16
Name: price, dtype: str

In [36]:
possible_date_columns = [
    "last_scraped",
    "host_since",
    "calendar_last_scraped",
    "first_review",
    "last_review"
]

for column in possible_date_columns:
    if column in listings_detailed.columns:
        print(f"\n{column}")

        display(
            listings_detailed[column]
            .dropna()
            .head()
        )


last_scraped


0    2026-07-02
1    2026-07-02
2    2026-07-02
3    2026-07-02
4    2026-07-02
Name: last_scraped, dtype: str


host_since


Series([], Name: host_since, dtype: float64)


calendar_last_scraped


0    2026-07-02
1    2026-07-02
2    2026-07-02
3    2026-07-02
4    2026-07-02
Name: calendar_last_scraped, dtype: str


first_review


0    2011-12-19
1    2011-07-30
2    2011-05-04
3    2012-01-28
4    2012-03-13
Name: first_review, dtype: str


last_review


0    2020-01-17
1    2019-10-13
2    2020-01-09
3    2020-01-21
4    2023-08-01
Name: last_review, dtype: str

In [37]:
display(calendar["date"].head())

print(
    "Earliest calendar date:",
    calendar["date"].min()
)

print(
    "Latest calendar date:",
    calendar["date"].max()
)

0    2026-07-02
1    2026-07-03
2    2026-07-04
3    2026-07-05
4    2026-07-06
Name: date, dtype: str

Earliest calendar date: 2026-06-29
Latest calendar date: 2027-07-01


In [38]:
print(
    "Earliest review date:",
    reviews_detailed["date"].min()
)

print(
    "Latest review date:",
    reviews_detailed["date"].max()
)

Earliest review date: 2011-05-04
Latest review date: 2026-06-29


In [39]:
output_directory = (
    project_root /
    "outputs" /
    "data_quality"
)

output_directory.mkdir(
    parents=True,
    exist_ok=True
)

dataset_overview_df.to_csv(
    output_directory /
    "dataset_overview.csv",
    index=False
)

for dataset_name, profile in profiles.items():
    safe_name = (
        dataset_name.lower()
        .replace(" ", "_")
    )

    profile.to_csv(
        output_directory /
        f"{safe_name}_profile.csv",
        index=False
    )

duplicate_summary.to_csv(
    output_directory /
    "duplicate_summary.csv",
    index=False
)

print(
    "Profiling reports saved to:",
    output_directory
)

Profiling reports saved to: e:\PROJECTS (selfstudy)\expernetic-airbnb-data-engineering\outputs\data_quality


# Initial Dataset Findings

## Dataset coverage

The project uses the Singapore Inside Airbnb dataset dated 29 June 2026.
The downloaded files include detailed listings, calendar records, detailed
reviews, summary listings, summary reviews, neighbourhood reference data,
and neighbourhood GeoJSON boundaries.

## Dataset relationships

- `listings.id` links to `calendar.listing_id`.
- `listings.id` links to `reviews.listing_id`.
- `listings.host_id` represents the host entity.
- `listings.neighbourhood_cleansed` links listings to neighbourhoods.
- Calendar and review dates can later link to a date dimension.

## Initial data-quality observations

- Some listing and host fields contain missing values.
- Price fields are stored as text and require currency cleaning.
- Percentage fields require conversion from text to numeric values.
- Date fields require standard parsing.
- Availability values require conversion from `t` and `f` to Boolean values.
- Missing review scores should not be treated as negative ratings because some
  listings have not received reviews.

## Important limitations

Calendar availability does not confirm whether a listing was booked. An
unavailable date may represent a booking, a host block, maintenance, or
another reason. Therefore, any occupancy measure created from calendar data
will be described as an estimated occupancy proxy.

In [40]:
def create_id_set(
    dataframe: pd.DataFrame,
    column_name: str
) -> set[int]:
    return set(
        pd.to_numeric(
            dataframe[column_name],
            errors="coerce"
        )
        .dropna()
        .astype("int64")
    )


detailed_listing_ids = create_id_set(
    listings_detailed,
    "id"
)

summary_listing_ids = create_id_set(
    listings_summary,
    "id"
)

calendar_listing_ids = create_id_set(
    calendar,
    "listing_id"
)

detailed_review_listing_ids = create_id_set(
    reviews_detailed,
    "listing_id"
)

summary_review_listing_ids = create_id_set(
    reviews_summary,
    "listing_id"
)

In [41]:
summary_only_ids = (
    summary_listing_ids -
    detailed_listing_ids
)

detailed_only_ids = (
    detailed_listing_ids -
    summary_listing_ids
)

print(
    "Summary-only listing IDs:",
    len(summary_only_ids)
)

print(
    "Detailed-only listing IDs:",
    len(detailed_only_ids)
)

Summary-only listing IDs: 150
Detailed-only listing IDs: 0


In [42]:
relationship_checks = pd.DataFrame([
    {
        "check": "Calendar IDs missing from detailed listings",
        "count": len(
            calendar_listing_ids -
            detailed_listing_ids
        )
    },
    {
        "check": "Calendar IDs missing from summary listings",
        "count": len(
            calendar_listing_ids -
            summary_listing_ids
        )
    },
    {
        "check": "Review IDs missing from detailed listings",
        "count": len(
            detailed_review_listing_ids -
            detailed_listing_ids
        )
    },
    {
        "check": "Review IDs missing from summary listings",
        "count": len(
            detailed_review_listing_ids -
            summary_listing_ids
        )
    }
])

display(relationship_checks)

,check,count
0,Calendar IDs missing from detailed listings,150
1,Calendar IDs missing from summary listings,0
2,Review IDs missing from detailed listings,103
3,Review IDs missing from summary listings,0


In [43]:
review_ids_missing_detailed = (
    detailed_review_listing_ids -
    detailed_listing_ids
)

print(
    "Review IDs missing from detailed listings:",
    len(review_ids_missing_detailed)
)

print(
    "All missing review IDs exist in summary-only listings:",
    review_ids_missing_detailed.issubset(
        summary_only_ids
    )
)

Review IDs missing from detailed listings: 103
All missing review IDs exist in summary-only listings: True


In [44]:
calendar_days_per_listing = (
    calendar
    .groupby("listing_id")["date"]
    .nunique()
)

display(
    calendar_days_per_listing
    .describe()
)

count    3247.0
mean      365.0
std         0.0
min       365.0
25%       365.0
50%       365.0
75%       365.0
max       365.0
Name: date, dtype: float64

In [45]:
print(
    "Listings with exactly 365 calendar dates:",
    (calendar_days_per_listing == 365).sum()
)

print(
    "Listings with fewer than 365 calendar dates:",
    (calendar_days_per_listing < 365).sum()
)

print(
    "Listings with more than 365 calendar dates:",
    (calendar_days_per_listing > 365).sum()
)

Listings with exactly 365 calendar dates: 3247
Listings with fewer than 365 calendar dates: 0
Listings with more than 365 calendar dates: 0


In [46]:
review_counts_by_listing_date = (
    reviews_detailed
    .groupby(
        ["listing_id", "date"]
    )["id"]
    .nunique()
    .reset_index(
        name="review_count"
    )
)

multiple_reviews_same_day = (
    review_counts_by_listing_date[
        review_counts_by_listing_date[
            "review_count"
        ] > 1
    ]
)

extra_same_day_reviews = (
    multiple_reviews_same_day[
        "review_count"
    ] - 1
).sum()

print(
    "Listing-date combinations with multiple reviews:",
    len(multiple_reviews_same_day)
)

print(
    "Additional reviews sharing a listing and date:",
    extra_same_day_reviews
)

display(
    multiple_reviews_same_day
    .sort_values(
        "review_count",
        ascending=False
    )
    .head(10)
)

Listing-date combinations with multiple reviews: 2414
Additional reviews sharing a listing and date: 3165


,listing_id,date,review_count
35308,1394962520701277726,2026-03-20,9
35347,1394962520701277726,2026-05-03,9
35547,1394962792824689414,2026-05-18,8
35346,1394962520701277726,2026-05-02,8
35350,1394962520701277726,2026-05-07,8
35281,1394962520701277726,2026-02-19,7
35341,1394962520701277726,2026-04-26,7
35253,1394962520701277726,2026-01-17,7
35295,1394962520701277726,2026-03-07,6
35381,1394962520701277726,2026-06-13,6


In [47]:
relationship_checks.to_csv(
    output_directory /
    "relationship_checks.csv",
    index=False
)

pd.DataFrame({
    "listing_id": sorted(
        summary_only_ids
    )
}).to_csv(
    output_directory /
    "summary_only_listing_ids.csv",
    index=False
)

pd.DataFrame({
    "listing_id": sorted(
        review_ids_missing_detailed
    )
}).to_csv(
    output_directory /
    "review_ids_missing_detailed.csv",
    index=False
)

print("Relationship outputs saved successfully.")

Relationship outputs saved successfully.


In [48]:
summary_only_listings = (
    listings_summary[
        listings_summary["id"].isin(
            summary_only_ids
        )
    ]
    .copy()
)

print(
    "Summary-only rows:",
    len(summary_only_listings)
)

display(
    summary_only_listings.head(10)
)

Summary-only rows: 150


,id,name,host_id,host_profile_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365,number_of_reviews_ltm,license
2216,1289634609448372677,Habyt Kada - Deluxe Queen Room,NaN,NaN,NaN,Central Region,Downtown Core,1.280558,103.845189,Private room,168.0,1,1,2025-03-27,0.06,NaN,252,0,NaN
2217,1289634609559578669,Habyt Kada - Studio Room,NaN,NaN,NaN,Central Region,Downtown Core,1.280558,103.845189,Private room,168.0,1,0,NaN,NaN,NaN,132,0,NaN
2291,1315652008053341890,Deluxe Studio no window @ Habyt Studios Kallang,NaN,NaN,NaN,Central Region,Kallang,1.311059,103.873765,Private room,187.0,1,0,NaN,NaN,NaN,345,0,NaN
2294,1315652094356442416,"Studio Loft, spacious and comfortable",NaN,NaN,NaN,Central Region,Kallang,1.311059,103.873765,Private room,187.0,1,2,2025-08-14,0.15,NaN,305,1,NaN
2295,1315652514612739647,Superior studio single,NaN,NaN,NaN,Central Region,Novena,1.326341,103.847740,Private room,145.0,1,2,2025-04-03,0.12,NaN,348,0,NaN
2296,1315652515073850821,Superior studio queen (no window),NaN,NaN,NaN,Central Region,Novena,1.326341,103.847740,Private room,155.0,1,2,2025-10-22,0.13,NaN,351,1,NaN
2299,1315652688630554864,Premier studio queen,NaN,NaN,NaN,Central Region,Novena,1.326341,103.847740,Private room,135.0,1,2,2026-03-03,0.17,NaN,349,2,NaN
2388,1360442149406882719,Comfort Deluxe Queen Room with Street View,NaN,NaN,NaN,Central Region,Rochor,1.304930,103.853650,Private room,170.0,1,2,2025-05-20,0.13,NaN,361,0,NaN
2389,1360443252300949053,Comfort Deluxe Twin Room with Street View,NaN,NaN,NaN,Central Region,Rochor,1.304930,103.853650,Private room,170.0,1,1,2025-03-29,0.07,NaN,362,0,NaN
2423,1368584855837908377,"Hotel Classic by Venue, Standard queen (no win...",NaN,NaN,NaN,Central Region,Geylang,1.315720,103.897706,Private room,121.0,1,55,2026-06-14,4.17,NaN,336,52,NaN


In [49]:
columns_to_view = [
    "id",
    "name",
    "host_id",
    "host_name",
    "neighbourhood",
    "latitude",
    "longitude",
    "room_type",
    "price",
    "number_of_reviews",
    "availability_365"
]

available_columns = [
    column
    for column in columns_to_view
    if column in summary_only_listings.columns
]

display(
    summary_only_listings[
        available_columns
    ].head(20)
)

,id,name,host_id,host_name,neighbourhood,latitude,longitude,room_type,price,number_of_reviews,availability_365
2216,1289634609448372677,Habyt Kada - Deluxe Queen Room,NaN,NaN,Downtown Core,1.280558,103.845189,Private room,168.0,1,252
2217,1289634609559578669,Habyt Kada - Studio Room,NaN,NaN,Downtown Core,1.280558,103.845189,Private room,168.0,0,132
2291,1315652008053341890,Deluxe Studio no window @ Habyt Studios Kallang,NaN,NaN,Kallang,1.311059,103.873765,Private room,187.0,0,345
2294,1315652094356442416,"Studio Loft, spacious and comfortable",NaN,NaN,Kallang,1.311059,103.873765,Private room,187.0,2,305
2295,1315652514612739647,Superior studio single,NaN,NaN,Novena,1.326341,103.847740,Private room,145.0,2,348
2296,1315652515073850821,Superior studio queen (no window),NaN,NaN,Novena,1.326341,103.847740,Private room,155.0,2,351
2299,1315652688630554864,Premier studio queen,NaN,NaN,Novena,1.326341,103.847740,Private room,135.0,2,349
2388,1360442149406882719,Comfort Deluxe Queen Room with Street View,NaN,NaN,Rochor,1.304930,103.853650,Private room,170.0,2,361
2389,1360443252300949053,Comfort Deluxe Twin Room with Street View,NaN,NaN,Rochor,1.304930,103.853650,Private room,170.0,1,362
2423,1368584855837908377,"Hotel Classic by Venue, Standard queen (no win...",NaN,NaN,Geylang,1.315720,103.897706,Private room,121.0,55,336


# Dataset Relationship Findings

The summary listings file contains 3,247 unique listings, while the detailed
listings file contains 3,097. Therefore, 150 listings appear only in the
summary file.

The calendar contains 1,185,155 records, which is exactly 3,247 listings
multiplied by 365 calendar dates. This indicates that the calendar dataset
uses the summary listings file as its complete listing population.

A total of 150 calendar listing IDs and 103 review listing IDs are absent
from the detailed listings file. These records are not automatically treated
as invalid because they are present in the summary listings dataset.

The detailed reviews file and summary reviews file both contain 41,265 rows.
The apparent duplicates in the summary file occur because it contains only
listing ID and review date. Multiple distinct reviews may therefore appear
identical when they were submitted for the same listing on the same date.

The detailed reviews dataset will be used as the primary review source
because it contains a unique review ID and additional review attributes.

In [51]:
from pathlib import Path

import pandas as pd


if Path.cwd().name == "notebooks":
    project_root = Path.cwd().parent
else:
    project_root = Path.cwd()

validation_report = pd.read_csv(
    project_root
    / "outputs"
    / "data_quality"
    / "validation_report.csv"
)

failed_rules = validation_report[
    validation_report["failed_rows"] > 0
].copy()

display(
    failed_rules[
        [
            "dataset",
            "validation_rule",
            "failed_rows",
            "failed_percentage"
        ]
    ]
)

,dataset,validation_rule,failed_rows,failed_percentage
2,listing_master,missing_price,505,15.5528


In [54]:
import pyarrow.parquet as pq

invalid_path = (
    project_root
    / "data"
    / "quarantine"
    / "listing_master_invalid.parquet"
)

arrow_table = pq.read_table(
    invalid_path,
    arrow_extensions_enabled=False
)

invalid_listings = arrow_table.to_pandas()

display(
    invalid_listings[
        "validation_errors"
    ].value_counts(
        dropna=False
    )
)

Series([], Name: count, dtype: Int64)

In [55]:
import duckdb


database_path = (
    project_root
    / "outputs"
    / "database"
    / "airbnb_singapore.duckdb"
)

connection = duckdb.connect(
    str(database_path),
    read_only=True
)

In [56]:
tables = connection.execute(
    """
    SHOW TABLES
    """
).fetchdf()

display(tables)

,name
0,dim_date
1,dim_host
2,dim_listing
3,dim_neighbourhood
4,fact_calendar
5,fact_listing_snapshot
6,fact_reviews
7,pipeline_table_metadata
8,stg_calendar
9,stg_listing_enriched


In [57]:
display(
    connection.execute(
        """
        SELECT *
        FROM dim_listing
        LIMIT 10
        """
    ).fetchdf()
)

,listing_key,listing_id,host_key,neighbourhood_key,listing_name,property_type,room_type,accommodates,bedrooms,beds,latitude,longitude,record_source,has_detailed_record
0,1,71609,3,36,Ensuite Room (Room 1 & 2) near EXPO,Private room in villa,Private room,2.0,2.0,NaN,1.34537,103.95887,detailed,True
1,2,71896,3,36,B&B Room 1 near Airport & EXPO,Private room in home,Private room,1.0,NaN,NaN,1.34754,103.95958,detailed,True
2,3,71903,3,36,Room 2-near Airport & EXPO,Private room in home,Private room,2.0,NaN,NaN,1.34531,103.96100,detailed,True
3,4,294281,12,21,5 mins walk from Newton subway,Private room in rental unit,Private room,2.0,2.0,NaN,1.31142,103.83924,detailed,True
4,5,344803,3,36,Budget short stay room near EXPO,Private room in home,Private room,1.0,NaN,NaN,1.34917,103.96090,detailed,True
5,6,369141,12,21,5mins from Newton Train Station,Private room,Private room,2.0,NaN,NaN,1.31144,103.83784,detailed,True
6,7,369145,12,21,5 mins walk from Newton MRT,Private room in townhouse,Private room,2.0,2.0,NaN,1.31129,103.83788,detailed,True
7,8,395191,17,28,Heritage Apartment Room near Orchard,Private room in rental unit,Private room,1.0,NaN,NaN,1.29496,103.83575,detailed,True
8,9,468782,21,32,Lovely Ensuite Room in Serangoon,Private room in home,Private room,2.0,1.0,NaN,1.36288,103.86575,detailed,True
9,10,696136,10,7,Quiet ensuite room in Holland Village,Private room in home,Private room,2.0,NaN,NaN,1.30938,103.79812,detailed,True


In [58]:
display(
    connection.execute(
        """
        SELECT *
        FROM fact_listing_snapshot
        LIMIT 10
        """
    ).fetchdf()
)

,listing_key,snapshot_date_key,price,minimum_nights,maximum_nights,number_of_reviews,review_scores_rating,availability_30,availability_60,availability_90,availability_365,estimated_occupancy_proxy,estimated_availability_rate,detailed_review_count,host_tenure_years,price_per_bedroom,price_per_guest,reviews_per_host_year
0,1,20260702,NaN,92,1125.0,19,4.44,29.0,59.0,89.0,90,0.753425,0.246575,19,NaN,NaN,NaN,NaN
1,2,20260702,NaN,92,1125.0,24,4.16,29.0,59.0,89.0,90,0.753425,0.246575,24,NaN,NaN,NaN,NaN
2,3,20260702,NaN,92,1125.0,46,4.41,29.0,59.0,89.0,90,0.753425,0.246575,46,NaN,NaN,NaN,NaN
3,4,20260702,31.71,92,1125.0,131,4.43,29.0,59.0,89.0,364,0.002740,0.997260,131,NaN,15.855,15.855,NaN
4,5,20260702,14.66,92,1125.0,60,4.60,29.0,59.0,89.0,364,0.002740,0.997260,60,NaN,NaN,14.660,NaN
5,6,20260702,NaN,92,180.0,81,4.43,29.0,59.0,89.0,364,0.002740,0.997260,81,NaN,NaN,NaN,NaN
6,7,20260702,26.24,92,1125.0,163,4.39,29.0,59.0,89.0,364,0.002740,0.997260,163,NaN,13.120,13.120,NaN
7,8,20260702,NaN,93,365.0,27,4.31,22.0,52.0,82.0,83,0.772603,0.227397,27,NaN,NaN,NaN,NaN
8,9,20260702,17.67,92,180.0,83,4.70,29.0,59.0,89.0,364,0.002740,0.997260,83,NaN,17.670,8.835,NaN
9,10,20260702,NaN,92,1125.0,26,4.50,29.0,59.0,89.0,90,0.753425,0.246575,26,NaN,NaN,NaN,NaN


In [59]:
connection.close()

In [60]:
sql_output_directory = (
    project_root
    / "outputs"
    / "tables"
    / "sql_analysis"
)

market_overview = pd.read_csv(
    sql_output_directory
    / "market_overview.csv"
)

room_type_performance = pd.read_csv(
    sql_output_directory
    / "room_type_performance.csv"
)

neighbourhood_market_summary = pd.read_csv(
    sql_output_directory
    / "neighbourhood_market_summary.csv"
)

host_segment_summary = pd.read_csv(
    sql_output_directory
    / "host_segment_summary.csv"
)

market_concentration = pd.read_csv(
    sql_output_directory
    / "host_market_concentration.csv"
)

display(market_overview)
display(room_type_performance)
display(neighbourhood_market_summary.head(10))
display(host_segment_summary)
display(market_concentration)

,total_listings,total_hosts,total_neighbourhoods,median_listing_price,average_listing_price,average_review_rating,average_occupancy_proxy_percentage
0,3247,762,42,130.05,204.24,4.54,22.67


,room_type,listing_count,listing_share_percentage,median_price,average_price,average_rating,average_occupancy_proxy_percentage
0,Private room,1870,57.59,93.87,134.12,4.49,25.92
1,Entire home/apt,1226,37.76,214.11,312.56,4.66,18.11
2,Hotel room,83,2.56,117.20,165.55,4.39,19.78
3,Shared room,68,2.09,48.00,60.77,4.48,18.94


,neighbourhood_name,listing_count,unique_host_count,median_price,average_price,average_rating,average_occupancy_proxy_percentage,entire_home_share_percentage
0,Orchard,86,18,422.00,456.00,4.80,21.40,61.63
1,Southern Islands,21,9,421.00,639.95,4.82,21.55,57.14
2,Museum,18,12,350.50,383.22,4.33,15.72,38.89
3,River Valley,404,35,230.37,273.78,4.48,14.14,73.02
4,Tanglin,40,18,214.60,294.18,4.72,17.03,65.00
5,Novena,197,44,209.35,200.58,4.70,23.79,74.11
6,Singapore River,145,25,184.89,270.47,4.61,19.81,37.93
7,Hougang,44,25,162.40,263.24,4.69,28.29,47.73
8,Rochor,283,55,155.00,326.65,4.43,23.80,23.67
9,Queenstown,162,35,153.89,195.48,4.49,31.03,41.98


,host_segment,host_count,listing_count,listing_share_percentage,median_price,average_rating,average_occupancy_proxy_percentage
0,Large Operator,19,1279,39.39,162.69,4.59,12.70
1,Medium Portfolio,79,721,22.21,158.20,4.51,23.78
2,Small Portfolio,216,649,19.99,89.60,4.47,33.59
3,Single Listing,448,448,13.80,43.76,4.60,34.42
4,Unknown,0,150,4.62,152.50,NaN,19.99


,total_hosts,total_host_linked_listings,hosts_in_top_10_percent,listings_controlled_by_top_10_percent,top_10_percent_listing_share
0,762,3097.0,77,1874.0,60.51


In [61]:
execution_report = pd.read_csv(
    sql_output_directory
    / "sql_execution_report.csv"
)

display(execution_report)

print(
    "Successful queries:",
    (
        execution_report["status"]
        == "success"
    ).sum()
)

print(
    "Failed queries:",
    (
        execution_report["status"]
        == "failed"
    ).sum()
)

,query_name,status,output_rows,output_file,error_message
0,market_overview,success,1,outputs\tables\sql_analysis\market_overview.csv,NaN
1,room_type_performance,success,4,outputs\tables\sql_analysis\room_type_performa...,NaN
2,neighbourhood_market_summary,success,35,outputs\tables\sql_analysis\neighbourhood_mark...,NaN
3,host_segment_summary,success,5,outputs\tables\sql_analysis\host_segment_summa...,NaN
4,top_hosts_by_listing_count,success,20,outputs\tables\sql_analysis\top_hosts_by_listi...,NaN
5,host_market_concentration,success,1,outputs\tables\sql_analysis\host_market_concen...,NaN
6,superhost_comparison,success,3,outputs\tables\sql_analysis\superhost_comparis...,NaN
7,price_by_review_volume,success,5,outputs\tables\sql_analysis\price_by_review_vo...,NaN
8,monthly_review_activity,success,180,outputs\tables\sql_analysis\monthly_review_act...,NaN
9,monthly_availability,success,14,outputs\tables\sql_analysis\monthly_availabili...,NaN


Successful queries: 12
Failed queries: 0
